File: cpp/025.cpp
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *dummy = new ListNode(0);
        ListNode *tail = dummy;
        while (head != nullptr) {
            ListNode *kth = findKth(head, k);
            if (kth == nullptr) {
                tail->next = head;
                return dummy->next;
            }
            ListNode *temp = head;
            ListNode *next = kth->next;
            kth->next = nullptr;
            for (int i = 0; i < k; ++i) {
                ListNode *tmp = head->next;
                head->next = tail->next;
                tail->next = head;
                head = tmp;
            }
            tail = temp;
            head = next;
        }
        return dummy->next;
    }
    
    ListNode* findKth(ListNode* head, int k) {
        for (int i = 1; i < k && head != nullptr; ++i)
            head = head->next;
        return head;
    }
};

File: go/025.go
func reverseKGroup(head *ListNode, k int) *ListNode {
    if k <= 1 {
        return head
    }
    dummy := &ListNode{}
    tail := dummy
    for head != nil {
        kth := findKth(head, k)
        if kth == nil { //length less than k
            tail.Next = head
            break
        }
        next := kth.Next
        kth.Next = nil
        temp := head
        for i := 0; i < k; i++ {
            tmp := head.Next
            head.Next = tail.Next
            tail.Next = head
            head = tmp
        }
        tail = temp
        head = next
    }
    return dummy.Next
}

//return the Kth node, if list length less than k, then return nil
func findKth(head *ListNode, k int) *ListNode {
    for i := 1; i < k && head != nil; i++ {
        head = head.Next
    }
    if head != nil {
        return head
    }
    return nil
}
